### PR TITLE
Trees: define KwErased, Mod.Erased, Type.FuncParam

### PR DESCRIPTION
**Pull request by kitbellew**
For #2774.

### CODE REVIEW

File: scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala [hunk 1]
```diff
@@ -367,7 +367,13 @@ object Type {
     checkParent(ParentChecks.TypeVar)
   }
 
-  @ast class TypedParam(name: Name, typ: Type) extends Type with Member.Type
+  @ast class TypedParam(name: Name, typ: Type, @newField("4.7.8") mods: List[Mod] = Nil)
+      extends Type with Member.Type with Stat.WithMods
+
+  @ast class FuncParam(mods: List[Mod], tpe: Type) extends Type with Stat.WithMods {
```
Comment (tgodzik) [inline scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:373]:
```suggestion
  @ast class FunctionParam(mods: List[Mod], tpe: Type) extends Type with Stat.WithMods {
```
it's easier to read full names I think

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala [hunk 2]
```diff
@@ -695,7 +695,8 @@ object TreeSyntax {
         }
       case t: Type.ByName => m(ParamTyp, s(kw("=>"), " ", p(Typ, t.tpe)))
       case t: Type.Var => m(SimpleTyp, s(t.name.value))
-      case t: Type.TypedParam => m(SimpleTyp, s(t.name.value), ": ", p(Typ, t.typ))
+      case t: Type.FuncParam => m(ParamTyp, w(t.mods, " "), p(Typ, t.tpe))
```
Comment (tgodzik) [inline scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:698]: Are there cases that cover both `TypedParam` and `FuncParam` with erased keyword?

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala (hunk 2 above)
Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:698]: I meant test cases, sorry :sweat:

Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:698]: actually, I think there are some removed files for https://github.com/scalameta/scalameta/actions/runs/5313174771/jobs/9618700633?pr=3189 that we could unignore.

File: scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala (hunk 1 above)
Comment (kitbellew) [reply scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala]: done

File: scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala (hunk 2 above)
Comment (kitbellew) [reply scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala]: added test cases in SyntacticSuite. others are in ErasedDefsSuite and will change as the parser starts supporting.

**Comment by kitbellew** [comment]
@tgodzik all done, tests pass.

File: tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala [hunk 3]
```diff
@@ -1928,6 +1928,28 @@ class SyntacticSuite extends scala.meta.tests.parsers.ParseSuite {
...
+    val tree = Type.Function(
+      List(Type.TypedParam(Type.Name("e"), Type.Name("Entry"), List(Mod.Erased()))),
+      Type.Select(Term.Name("e"), Type.Name("Key"))
+    )
+    assertSyntax(tree, "(erased e: Entry) => e.Key")(tree)
+  }
+
+  test("#2774 3") {
+    val tree = Type.Function(
+      List(Type.FunctionArg(List(Mod.Erased()), Type.Name("Ev")), Type.Name("Int")),
```
Comment (tgodzik) [inline tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala:1947]: Actually do we even need `FunctionArg` Shouldn't it be List(Type.TypedParam(Type.Name.Anonymous, Type.Name("Entry"), List(Mod.Erased()))), maybe?

Feels like `Type.FunctionArg` is just for mods, which is a bit of a inconsistency :thinking:

Comment (tgodzik) [reply tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala:1947]: Ach, that's fine then. Thanks for explaning!

**Pull request merged**